a2c/a2c.py

```diff
@@ -119,27 +102,54 @@ def train(self):
                                  np_entropy)
 
             current_state = observation
-            if done or simulation_step == self.batch_size - 1:
-                _, Qval = self.net.forward(observation)
+            if done or simulation_step == batch_size - 1:
+                Qval, _ = self.net.forward(observation)
                 np_Qval = Qval.detach().numpy()[0, 0]
                 break
 
         self.calculate_qvals(np_Qval)
 
-        tensor_estim_values = torch.FloatTensor(self.np_estim_values)
-        tensor_Qvals = torch.FloatTensor(self.np_Qvals)
 
+class A2CTrainer:
+    def __init__(self, net, out_num, environment_name, batch_size,
+                 gamma, beta_entropy, learning_rate, clip_size):
+
+        self.net = net
+        self.batch_size = batch_size
+        self.beta_entropy = beta_entropy
+        self.learning_rate = learning_rate
+        self.clip_size = clip_size
+        self.optimizer = torch.optim.Adam(
+            self.net.parameters(), lr=self.learning_rate)
+        self.data = DataCollector(net, out_num, environment_name, gamma)
+
+    def calculate_actor_loss(self, advantage, tensor_action_logarithms):
+        single_tensor_action_logarithms = torch.stack(
+            tensor_action_logarithms)
+        return (-single_tensor_action_logarithms * advantage).mean()
+
+    def calculate_critic_loss(self, advantage):
+        return 0.5 * advantage.pow(2).mean()
+
+    def train(self):
+        self.data.clear_previous_batch_data()
+
+        self.data.collect_data_for(batch_size=self.batch_size)
+
```

Comment:
No całkiem spoko podejście, oszczędza bałaganu

Reply:
Zobaczysz zaraz jak to się rozwinęło dalej.